### ISSUE TITLE
vsxmake generator: Add a BuildAll target

### ISSUE THREAD

**Issue by SirLynix** [xmake-io/xmake#2969]
### Is your feature request related to a problem? Please describe.

Hello,

When using a visual studio project generated by the vsxmake generator, clicking "Solution => Build all" (or worse "Solution => Rebuild all") will trigger `xmake build` / `xmake build -r` on every target.

This is an issue when you have several projects (targets), as multiple xmake instance will be spawned and will try to rebuild the project from scratch sequentially, meaning a lot of files will be rebuilt multiple times.

This isn't great as one of the benefit we have from using vsxmake is that xmake compiles faster than msbuild.

### Describe the solution you'd like

It's possible to tell Visual Studio to not build a project when building the whole solution, by removing the `{#target_id#}.#mode#|#arch#.Build.0 = #mode#|#vsarch#` association in the .sln file (this still allows to build a target but won't select it when building the solution).

It would be nice to have a solution similar to what cmake does: adding a phony target which builds everything (and will be the only one target with a Build.0 association), and use it to simply run `xmake build` or `xmake build -r` on the whole project.

Example:
![image](https://user-images.githubusercontent.com/3002461/197508734-4c564a82-2afa-459f-86ef-a5767da18886.png)

I began working on this as I wanted to make a pull request for this: https://github.com/xmake-io/xmake/commit/36a8591345fdff92395b53bef2fbbe46e10992b3  
however I wasn't sure of the best way to generate the BuildAll.vcxproj using the vsxmake generator.

Also, it would be really cool to have a .xmake/UpdateVS.vcxproj which would trigger the `xmake project -k vsxmake` manually, a bit like `plugin.vsxmake.autoupdate` rule but in a way we could trigger manually.

### Describe alternatives you've considered

There's no other solution than changing the way xmake generates project, we can use vs to tell it to not build projects when building the whole solution but xmake overrides this when it regenerates the .sln

### Additional context

I'm using last xmake version

**Comment by waruqi** [comment]
But I don't know how best to implement it, perhaps you could refer to the cmake generator.

**Comment by SirLynix** [comment]
It's should be easy to implement, it's like a regular target which will trigger xmake, but instead of triggering `xmake build TargetName` it should just call `xmake build`.

My only difficulty is to generate this file with the current .vcxproj generation.

**Comment by waruqi** [comment]
maybe you just need add a phony target and disable it by default.

```lua
rule("plugin.vsxmake.buildall")
    on_buildcmd(function(batchcmds)
        -- add build all command
    end)
target("BuildAll")
    set_kind("phony")
    set_default(false)
    add_rules("plugin.vsxmake.buildall")
```

and UpdateVS too.

**Comment by waruqi** [comment]
> I though of that, but that won't prevent visual studio from starting two xmake process per targets when "build solution" is triggered.
> 
> I think it would be best to have it as a part of the vsxmake generator, as it's a common need.

no, use rule/batchcmds will generate only one xmake process when you click BuildAll.

It just add custom command in vcproj.

**Comment by SirLynix** [comment]
Yes, but with that the "build solution" button won't have a correct behavior, which means the shortcut control+shift+B visual studio users are used to will rebuild a lot of files multiple times. This is the main issue which cannot be fixed by a user project.

**Comment by waruqi** [comment]
> Yes, but with that the "build solution" button won't have a correct behavior, which means the shortcut control+shift+B visual studio users are used to will rebuild a lot of files multiple times. This is the main issue which cannot be fixed by a user project.

? if we use BuildAll target + rule/batchcmds, and add `xmake build --all` custom command in vcproj, It will only build incrementally, not rebuild a lot files when we click `BuildAll`

**Comment by SirLynix** [comment]
It will build correctly if you build that "BuildAll" target, yes.

But the issue I'm talking about is using the "Build solution" button:
![](https://learn.microsoft.com/fr-fr/visualstudio/ide/media/vs-2022/build-tutorial-build-menu.png?view=vs-2022).

With the current vsxmake generator this will try to build every single project, starting two xmake process for each which will conflict and wait for each other, and build files in a non-optimal order or rebuild every file multiple times when using "Rebuild solution".

**Comment by waruqi** [comment]
oh, got it.

**Comment by SirLynix** [comment]
I will try to implement this.

I was also thinking of reworking vsxmake projects to use "Makefile" type which allows to override Visual Studio behavior when compiling, for example:
```xml
<PropertyGroup>
  <NMakeBuildCommandLine>$(_XmakeExecutable) config $(_XmakeCommonFlags) $(_XmakeConfigFlags) && $(_XmakeExecutable) build $(_XmakeCommonFlags) $(_XmakeBuildFlags) $(_XmakeTarget)</NMakeBuildCommandLine>
  <NMakeReBuildCommandLine>$(_XmakeExecutable) config $(_XmakeCommonFlags) $(_XmakeConfigFlags) && $(_XmakeExecutable) build -r $(_XmakeCommonFlags) $(_XmakeBuildFlags) $(_XmakeTarget)</NMakeReBuildCommandLine>
  <NMakeCleanCommandLine>$(_XmakeExecutable) config $(_XmakeCommonFlags) $(_XmakeConfigFlags) && $(_XmakeExecutable) clean $(_XmakeCommonFlags) $(_XmakeCleanFlags) $(_XmakeTarget)</NMakeCleanCommandLine>
  <NMakeOutput>$(OutDir)$(TargetName)$(TargetExt)</NMakeOutput>
</PropertyGroup>
```

This is similar to what Unreal Engine does (since it uses a custom build tool), what do you think?

**Comment by waruqi** [comment]
NMakeBuildCommandLine will override whole vs build command?

**Comment by SirLynix** [comment]
Yes, when the project has type "Makefile" it expects you to give build commands and bypasses normal VS behavior.

It means all projects use the "Makefile" configuration type instead of binary/shared/static but those are not used anyway since xmake handles compilation.

https://learn.microsoft.com/en-us/cpp/build/reference/nmake-property-page?view=msvc-170

**Comment by waruqi** [comment]
This seems to be a nice feature. you can try it.

**Comment by Domain** [comment]
One simple solution is: set all other projects to non-default target, only BuildAll project is the default target. Then build solution will build the BuildAll project only. But I cannot call xmake again inside on_buildcmd, no matter what I used: os.run, task.run, process.open， all failed.

> maybe you just need add a phony target and disable it by default.
> 
> ```lua
> rule("plugin.vsxmake.buildall")
>     on_buildcmd(function(batchcmds)
>         -- add build all command
>     end)
> target("BuildAll")
>     set_kind("phony")
>     set_default(false)
>     add_rules("plugin.vsxmake.buildall")
> ```
> 
> and UpdateVS too. `plugin.vsxmake.update`
> 
> we can add `plugin.vsxmake.buildall` as builtin rule.
